src/realm/sync/client_syncsocket.hpp

```diff
@@ -0,0 +1,235 @@
+///////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2022 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_SYNC_CLIENT_SYNC_SOCKET_HPP
+#define REALM_SYNC_CLIENT_SYNC_SOCKET_HPP
+
+#include <map>
+#include <memory>
+#include <string>
+
+#include <realm/status.hpp>
+
+#include <realm/sync/config.hpp>
+
+#include <realm/util/functional.hpp>
+#include <realm/util/optional.hpp>
+
+namespace realm::sync {
+
+/// Struct that defines the endpoint to create a new websocket connection.
+/// Many of these values come from the SyncClientConfig passed to SyncManager when
+/// it was created.
+struct Endpoint {
+    using port_type = sync::port_type;
+
+    std::string address;   // Host address
+    port_type port;        // Host port number
+    std::string path;      // Includes access token in query.
+    std::string protocols; // One or more websocket protocols, separated with ", "
+    bool is_ssl;           // true if SSL should be used
+
+    /// The remaining fields are just pass through values from SyncConfig. They may
+    /// not be provided if the SDK chooses not to support the related config options.
+    /// These may be necessary when using websocket libraries without low-level
+    /// control. These items may be removed or converted to test only parameters
+    /// in the future.
+    std::map<std::string, std::string> headers; // Only includes "custom" headers.
```

Comment:
I kind of think all these items (line 52-56) should be in a struct passed into whatever factory creates the implementation of `WebSocketInterface`

So you'd create a DefaultWebSocketInterfaceFactory that takes these parameters, and then each call to connect() returns a websocket configured with these. Then for websocket implementations that don't support these or that support these natively don't have to pass them into realm at all.

Reply:
Yes - they don't change between websocket connections, so they can be added to the (new) DefaultSyncSocket constructor, and custom implementations can set them directly themselves when the SyncSocket shared_object is created and added to SyncClientConfig.